Reintroduce broadphase

jcarpent opened the pull request:
This requires C++11 and more.

jmirabel reviewed (commented):
Two questions:
- Did you make changes in the broadphase file that you re-introduce ?
- Is there anything in particular that requires special attention in those files ?

jcarpent commented on CMakeLists.txt:
```diff
@@ -164,6 +166,41 @@ SET(${PROJECT_NAME}_HEADERS
+  include/hpp/fcl/broadphase/broadphase_bruteforce.h
+  include/hpp/fcl/broadphase/broadphase_collision_manager.h
+  include/hpp/fcl/broadphase/broadphase_continuous_collision_manager-inl.h
+  include/hpp/fcl/broadphase/broadphase_continuous_collision_manager.h
```
I do agree. I put it knowing that it will need to do some reworking.

jcarpent commented on doc/python/doxygen_xml_parser.py:
```diff
@@ -164,7 +164,18 @@ def xmlToType (self, node, array=None, parentClass=None, tplargs=None):
+                    self_has_templates = c.tail is not None and c.tail.strip().find('<') != -1
+                    if self_has_templates:
+                        t += " template "
+                    t += c.text.strip() 
```
Yes, not easy to follow.

jcarpent commented on include/hpp/fcl/collision.h:
```diff
@@ -94,54 +94,24 @@ inline std::size_t collide(const CollisionGeometry* o1, const Transform3f& tf1,
 protected:
-  CollisionGeometry const *o1, *o2;
+  mutable const CollisionGeometry *o1;
+  mutable const CollisionGeometry *o2;
```
The idea is that we might need to update these values for external application like with Pinocchio when updating some geometries for instance.

jmirabel replied:
I think you should rather create a new `Compute...` object. This class caches the steps to identify which function to call and you shouldn't rely on the way this is done.

For instance, it may make sense to call a different function based on the number of triangles of a BVHModel. Changing the geometry may invalidate the cache.

jcarpent replied:
Yes, I know. This is just for the particular case where you keep the same geometry type, but you change the points of this geometry like updating an height field

jcarpent replied:
I think the addition of mutable is only for expert people like us I would say.

jmirabel replied:
I don't understand why you make it mutable in this class then. You can access the non-const model from the geometry vector in Pinocchio GeomModel class, can't you ?

jcarpent replied:
The thing is that when you erase the pointer of a given GeometryObject with a new one (because you load a new geometry of same type), then you might need to also update the value of the pointer within ComputeCollision functor classes. This is the purpose.

jcarpent commented:
> Two questions:
> 
>     * Did you make changes in the broadphase file that you re-introduce ?
>

I changed some stuff to make the API more flexible. For instance, I have introduced the notion of Callback classes to enable external applications to derive the operators in use for particular cases, such as with Pinocchio.
Compared to the original version, it is then more flexible and extendable I would say.

jcarpent commented:
> * Is there anything in particular that requires special attention in those files ?

I would say we should merge it, use it and extend it corresponding to our needs.
I did not change the underlying broadphase algorithms, even if some of them can be largely improved.

jcarpent commented on include/hpp/fcl/collision.h:
```diff
@@ -94,54 +94,24 @@ inline std::size_t collide(const CollisionGeometry* o1, const Transform3f& tf1,
-    std::size_t res = operator()(tf1, tf2, (const CollisionRequest&) request, result);
-    request.updateGuess (result);
-    return res;
-  }
```
So in the end, which signature do you want me to inline?

jmirabel replied:
The one which takes a non-const request.

jcarpent replied:
I will undo the commit. It will be simpler.

jcarpent commented:
@jmirabel I've handled your comments ;)

wxmerkt commented on CMakeLists.txt:
```diff
@@ -74,6 +74,9 @@ APPLY_DEFAULT_APPLE_CONFIGURATION()
 PROJECT(${PROJECT_NAME} ${PROJECT_ARGS})
 
+# Default C++ version should be C++11
+CHECK_MINIMAL_CXX_STANDARD(11 ENFORCE)
```
Should this be enforced to exactly 11 or at minimum 11? I.e. if a dependent code is on 14, this would change the variable CMAKE_CXX_STANDARD to exactly 11 if I recall the macro correctly?

wxmerkt replied:
I updated the comment - I mixed up the meaning of ENFORCE in the macro (it automatically updates only if lower than required, it won't downgrade or enforce a specific version)

proyan commented:
@nim65s Could we relaunch the gitlab-ci please? It seems to have failed because of the maintainence

jmirabel reviewed (approved):
Fine for me. Thanks for the work.

Maybe we should say a word on this in the acknowledgement section of the readme. At the moment, it is
```md
The development of **HPP-FCL** is actively supported by the [Gepetto team](http://projects.laas.fr/gepetto/) [@LAAS-CNRS](http://www.laas.fr) and the [Willow team](https://www.di.ens.fr/willow/) [@INRIA](http://www.inria.fr).
```
and could be
```md
[FCL](https://github.com/flexible-collision-library/fcl) was forked in 2015. Since then, a large part of the code has been rewritten or removed (for the unused part). The broadphase was reintroduced by J. Carpentier in 2022 based on the FCL version 0.7.0.

The development of **HPP-FCL** is actively supported by the [Gepetto team](http://projects.laas.fr/gepetto/) [@LAAS-CNRS](http://www.laas.fr), the [Willow team](https://www.di.ens.fr/willow/) [@INRIA](http://www.inria.fr) and, to some extend, [Eureka Robotics](https://eurekarobotics.com/).
```

nim65s commented:
> @nim65s Could we relaunch the gitlab-ci please? It seems to have failed because of the maintainence

done
